django_mongodb_backend/indexes.py

```diff
@@ -101,6 +105,191 @@ def where_node_idx(self, compiler, connection):
+                    "This MongoDB server does not support atlas search.",
+                    hint=(
+                        "The index won't be created. Silence this warning if you "
+                        "don't care about it."
+                    ),
+                    obj=self,
+                    id=f"{self._error_id_prefix}.W001",
+                )
+            )
+        return errors
+
+    # Maps Django internal type to atlas search index type.
+    # Reference: https://www.mongodb.com/docs/atlas/atlas-search/define-field-mappings/#data-types
+    def search_index_data_types(self, field, db_type):
+        if field.get_internal_type() == "UUIDField":
+            return "uuid"
+        if field.get_internal_type() in ("ObjectIdAutoField", "ObjectIdField"):
+            return "ObjectId"
+        if field.get_internal_type() == "EmbeddedModelField":
+            return "embeddedDocuments"
+        if db_type in ("int", "long"):
+            return "number"
+        if db_type == "binData":
+            return "string"
+        if db_type == "bool":
+            return "boolean"
+        if db_type == "object":
+            return "document"
+        return db_type
+
+    def get_pymongo_index_model(
+        self, model, schema_editor, field=None, unique=False, column_prefix=""
+    ):
+        if not schema_editor.connection.features.supports_atlas_search:
+            return None
+        fields = {}
+        for field_name, _ in self.fields_orders:
+            field_ = model._meta.get_field(field_name)
+            type_ = self.search_index_data_types(field_, field_.db_type(schema_editor.connection))
+            field_path = column_prefix + model._meta.get_field(field_name).column
+            fields[field_path] = {"type": type_}
+        return SearchIndexModel(
+            definition={"mappings": {"dynamic": False, "fields": fields}}, name=self.name
+        )
+
+
+class VectorSearchIndex(SearchIndex):
+    suffix = "vsi"
+    ALLOWED_SIMILARITY_FUNCTIONS = frozenset(("euclidean", "cosine", "dotProduct"))
+    _error_id_prefix = "django_mongodb_backend.indexes.VectorSearchIndex"
+
+    def __init__(self, *expressions, similarities="cosine", **kwargs):
+        super().__init__(*expressions, **kwargs)
+        self.similarities = similarities
+
+    def check(self, model, connection):
+        errors = super().check(model, connection)
+        similarities = (
+            self.similarities if isinstance(self.similarities, list) else [self.similarities]
+        )
+        for func in similarities:
+            if func not in self.ALLOWED_SIMILARITY_FUNCTIONS:
+                errors.append(
+                    Error(
+                        f"{func} isn't a valid similarity function, options "
+                        f"are {', '.join(sorted(self.ALLOWED_SIMILARITY_FUNCTIONS))}",
+                        obj=self,
+                        id=f"{self._error_id_prefix}.E004",
+                    )
+                )
```

Comment:
Anything that can be validated in `Index.__init__()` like similarities should be done there. That raises an error earlier in the process, especially because there's an issue in Django where `makemigrations` doesn't currently execute database checks like these. Validation that requires the model or a database connection can be left as a check.